google-cloud-bigquery/lib/google/cloud/bigquery/dataset.rb

```diff
@@ -2810,7 +2810,7 @@ def insert_async table_id, skip_invalid: nil, ignore_unknown: nil, max_bytes: 10
           ensure_service!
 
           # Get table, don't use Dataset#table which handles NotFoundError
-          gapi = service.get_table dataset_id, table_id, metadata_view: view
+          gapi = service.get_project_table project_id, dataset_id, table_id, metadata_view: view
```

Comment:
I don't understand how this changes things. Under what circumstance would the dataset's project ID be different from the project ID of the service backing the dataset?

Reply:
Oh, I see, it's now possible for the client (whose project is tied to the billing account) to get and construct a Dataset object from a different project. So this is actually a bug fix rather than a feature: in such a case, we're making sure the correct project gets set when doing an insert rows operation. Is that correct?

Reply:
> Oh, I see, it's now possible for the client (whose project is tied to the billing account) to get and construct a Dataset object from a different project. So this is actually a bug fix rather than a feature: in such a case, we're making sure the correct project gets set when doing an insert rows operation. Is that correct?

yeah, it might make sense to change it to a fix instead of feature.

Reply:
If this is indeed a bug fix (and I think it should be: I would expect a new feature to include an addition to the public interface, and this PR doesn't make any public interface changes at all), then I'd like to see new unit tests that show the fixed case. You can mock out the backend service, as many of the unit tests do, but just test that the correct project ID gets passed down to the backend service.